[Liu Haoran] ip

yuqiaoluolong opened the pull request:
Add some paragraph to the README.md to explain ...
Also add a heading ...

Commit:
Level-0: Greet

Commit:
Level-1: Greet, Echo, Exit
& a little bit customizes: 1. add online of " Here is yuqiaoluolong's Duke: " before each echo.
					2. add a couple of "<" and ">" in the beginning and the end of the horizontalLine

Commit:
Level-2: Add, List
&customizes:  1. add the second indentation
			2. add a line between Duke's response and user's input command
&optimizations:	1. delete unused import java.io.*;

Commit:
A-Classes: add markAsDone method

Commit:
Level-3: Mark as Done

Commit:
Naming: Some variables are named incorrectly and switch and while do not have correct spacing

It is messy if the naming and spacing does not follow some rules

Let's
*change String name Greet to greet
*change String name Bye to bye
*add a space between while and "("
*delete the redundant spacing before "case" in switch statement

Commit:
Level 4. ToDos, Events, Deadlines & A-Inheritance
original main classes: level-3
original Task classes: level-3

To achieve level-4, need to apply polymorphism
Task classes: add a method toString()
Todo classes: extends Task
Deadline classes: extends Task add one more variable by (ddl time)
Event classes: extends Task add one more variable at (happening time)
main class: edit the format of output according to level-4 requirements and add a condition when user typed unexpected strings

okkhoy commented:
If this is your submission for this week's IP task, please edit the title to follow the specified convention: https://nus-cs2113-ay2021s1.github.io/website/schedule/week4/project.html#1-create-a-pr-to-the-upstream-repo-mon-aug-31st-2359

@yuqiaoluolong for your action

Commit:
A-TextUiTesting

Edit the file called input.txt and Expected.TXT to test
There is still a problem of Troubleshooting and don't know how to use the dos2unix utility

Commit:
A-CodeQuality

Change the name of string respond to outputStatement
Add one String variable wrongWarning to hold a long string
Add two String variables description and date to hold the this.description and by or at date entered by user

yuqiaoluolong commented:
Thank you, prof. I have changed my PR title accordingly.

okkhoy commented:
Great! Thanks 🙂

chloesyy commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,92 @@
+    public static String tripleIndentation = doubleIndentation + "  ";
+    public static String horizontalLine = indentation
+            + "<------------------------------------------------------------>\n";
+    public static Task[] Tasks = new Task[100];
```
I think it would make the code more readable if you replace the 100 with a constant name, maybe "MAX_TASKS" to describe the number?

Also, I think Tasks is a variable - should it be named "tasks" instead? :)

yuqiaoluolong replied:
Oh, thank you! I didn't notice that. :D

chloesyy commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,92 @@
+                isBye = true;
+                break;
+            default:
+                if (inputCommand.contains("done")) {
+                    doneNum = Integer.parseInt(inputCommand.replace("done", " ").trim());
+                    Tasks[doneNum-1].markAsDone();
+                    outputStatement = doubleIndentation + "Nice! I've marked this task as done: \n"
+                            + tripleIndentation + Tasks[doneNum-1].toString() + "\n";
+                } else if (inputCommand.contains("todo")||inputCommand.contains("deadline")
+                        ||inputCommand.contains("event")){
+                    if (inputCommand.contains("todo")) {
+                        Tasks[numberOfTasks] = new Todo(inputCommand.replace("todo"," ").trim());
+                    } else if (inputCommand.contains("deadline")) {
+                        description = inputCommand.substring(inputCommand.indexOf("deadline")+8,
+                                inputCommand.indexOf("/")).trim();
+                        date = inputCommand.substring(inputCommand.indexOf("/by")+3).trim();
+                        Tasks[numberOfTasks] = new Deadline(description, date);
+
+                    } else if (inputCommand.contains("event")) {
+                        description = inputCommand.substring(inputCommand.indexOf("event")+5,
+                                inputCommand.indexOf("/")).trim();
+                        date = inputCommand.substring(inputCommand.indexOf("/at")+3).trim();
+                        Tasks[numberOfTasks] = new Event(description, date);
+                    }
+                    outputStatement = doubleIndentation + "Got it. I've added this task:\n"
+                            + tripleIndentation + Tasks[numberOfTasks].toString() + "\n" + doubleIndentation
+                            + "Now you have " + (numberOfTasks+1) + " tasks in the list.\n";
+                    numberOfTasks++;
```
Hi! I think you can try extracting the code fragments into different methods to increase readability? :)

yuqiaoluolong replied:
Hi, thank you! got it.

chloesyy reviewed (commented):
:)

thesamchris commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,92 @@
+        Scanner in = new Scanner(System.in);
+        while (!isBye) {
+            inputCommand = in.nextLine();
+            switch (inputCommand.trim()) {
```
i like how you used switch / case statements here 👍

yuqiaoluolong replied:
Thank you for appreciation. What do you mean by saying extract out individual cases? I did not get it

thesamchris reviewed (commented):
overall you did a great job haoran ! i like how you used switch/case statements . should the individual cases be extracted out ?

Reinbowl commented:
Overall your code mostly follows the coding standard except for certain parts where you left out a space between the ) and {

Here are some parts:
![image](https://user-images.githubusercontent.com/60181391/92072115-0f672b00-ede3-11ea-8bcf-0233cce65fa4.png)

![image](https://user-images.githubusercontent.com/60181391/92072177-31f94400-ede3-11ea-9868-3cb260805167.png)

![image](https://user-images.githubusercontent.com/60181391/92072196-42a9ba00-ede3-11ea-93df-1d0269a0a149.png)

Constants like this should be final and variable name all uppercase:
![image](https://user-images.githubusercontent.com/60181391/92072299-856b9200-ede3-11ea-8d79-34d4b1584bac.png)

Reinbowl reviewed (commented):
Overall your code mostly follows the coding standard except for certain parts where you left out a space between the ) and {

Here are some parts:
![image](https://user-images.githubusercontent.com/60181391/92072115-0f672b00-ede3-11ea-8bcf-0233cce65fa4.png)

![image](https://user-images.githubusercontent.com/60181391/92072177-31f94400-ede3-11ea-9868-3cb260805167.png)

![image](https://user-images.githubusercontent.com/60181391/92072196-42a9ba00-ede3-11ea-93df-1d0269a0a149.png)

Constants like this should be final and variable name all uppercase:
![image](https://user-images.githubusercontent.com/60181391/92072299-856b9200-ede3-11ea-8d79-34d4b1584bac.png)

chloesyy commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,92 @@
+import java.util.Scanner;
+
 public class Duke {
+    public static String indentation = "    ";
```
Hi! In this case since the indentations are constants, would it be better to name them as finals and in capital letters instead? e.g. "INDENTATION" 

((Hi I'm back to do more reviews))

yuqiaoluolong replied:
oh, thank you! I've corrected them. Hahhah

chloesyy commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,92 @@
+        String greet = indentation + " Hello! I'm Duke\n"
+                + indentation + " What can I do for you?\n";
+        String bye = indentation + " Bye. Hope to see you again soon!\n";
+        String wrongWarning = "Sorry, I don't know what's your command. Can you type again?\n";
```
Same thing here - since the wrong warning is a constant string, should it be named WRONG_WARNING instead?

chloesyy reviewed (commented):
Overall I think the code would be more readable if the codes in the switch cases are extracted out in different methods, like "executeTodoCommand" or something? Good job though! 👍 :D

Commit:
CodeQuality

correct some names of constants
add some methods to make the code inside main class clearer

Commit:
Level-5: Handle Errors

4 conditions of exceptions are added
the empty todo command exception is caught
the empty deadline command exception is caught
the empty event command exception is caught
the command out of the scope exception is caught

Commit:
Merge branch 'branch-Level-5'

Commit:
A-Packages: Organize into Packages & two more exceptions

Duke.java is put into the package Duke.command
Task.java, Todo.java, Deadline.java, and Event.java are put into the package Duke.task

The exceptions caused by the command "done" and "done x" where x is out of the boundary of the task list.

Commit:
add one more exception catching done commands exceeding total index limit of Tasks[]

Commit:
A-Collections

use an ArrayList<Task> to store the tasks instead of Task[ ]

Commit:
A-Collections
2nd version
use an ArrayList<Task> to store the tasks

Commit:
Level 6. Delete

Added the support for deleting tasks from the list.

Commit:
Level-7: Save

Add fileWriting method so that  whenever the task list changes, the tasks would be saved in the hard disk automatically

Commit:
Merge branch 'branch-Level-7'

# Conflicts:
#	src/main/java/Duke/command/Duke.java

Commit:
Addressed  conflicts between master and branch level-7

manaswini05 commented on src/main/java/Duke/command/Duke.java:
```diff
@@ -0,0 +1,205 @@
+package Duke.command;
+
+import Duke.task.*;
```
It is recommended to not do a wildcard import

manaswini05 commented on src/main/java/Duke/command/Duke.java:
```diff
@@ -0,0 +1,205 @@
+            System.out.println(DOUBLEINDENTATION + (i+1) + "." + tasks.get(i).toString());
+        }
+        System.out.println(HORIZONTALLINE);
+    }
+    public static void executeDoneCommand(String inputCommand) {
```
Would be good if you could add spacing of one line here

manaswini05 commented on src/main/java/Duke/command/Duke.java:
```diff
@@ -0,0 +1,205 @@
+        }
+    }
+
+    public static void executeListCommand(String inputcommanc) {
```
Consider camel case for string name and also the typo

manaswini05 commented on src/main/java/Duke/command/Duke.java:
```diff
@@ -0,0 +1,205 @@
+                    + TRIPLEINDENTATION + tasks.get(doneNum - 1).toString() + "\n");
+        } catch (NullPointerException e) {
+            printStatement(DOUBLEINDENTATION + "☹ OOPS!!! There is no task with such an index.\n");
+        }       //catch the command "done x" and x is out of the doundary of the task list
```
Recommended to add comments in a more formal way

manaswini05 commented on src/main/java/Duke/command/Duke.java:
```diff
@@ -0,0 +1,205 @@
+    }
+    public static boolean executeDeadlineCommand(ArrayList<Task> tasks, int numberOfTasks, String inputCommand) {
+        try {
+            String description = getDescriptiong("deadline", inputCommand);
```
Typo at Description - would be good if you could change it

manaswini05 commented on src/main/java/Duke/command/Duke.java:
```diff
@@ -0,0 +1,205 @@
+            inputCommand = in.nextLine();
+            isNewTask = inputCommand.contains("todo") || inputCommand.contains("deadline")
+                    || inputCommand.contains("event");
+            switch (inputCommand.trim()) {
```
ensure SLAP in switch-case as one of your conditions is very lengthy

manaswini05 commented on src/main/java/Duke/command/Duke.java:
```diff
@@ -0,0 +1,205 @@
+                } else if (inputCommand.contains("delete")) {
+                    executeDeleteCommand(inputCommand);
+                    write(numberOfTasks, file);
+                } else if (isNewTask) {
+                    if (inputCommand.contains("todo")) {
+                        try {
+                            executeTodoCommand(tasks, numberOfTasks, inputCommand);
+                        } catch (TodoNullException e){      //catch the empty todo command exception
+                            printStatement(DOUBLEINDENTATION +
+                                    "☹ OOPS!!! The description of a todo cannot be empty.\n");
+                            numberOfTasks--;    //to deal with "numberOfTasks++ below
+                        }
+                        write(numberOfTasks+1, file);
+                    } else if (inputCommand.contains("deadline")) {
+                        isCommandEmpty = executeDeadlineCommand(tasks, numberOfTasks, inputCommand);
+                        if(isCommandEmpty){
+                            numberOfTasks--;    //to deal with "numberOfTasks++ below
+                        }
+                        write(numberOfTasks+1, file);
```
Try to avoid deep nesting

manaswini05 reviewed (commented):
Overall, good work! :) Consider adding comments in recommended style

Commit:
changed the location of the text file storing the list.

Commit:
A-MoreOOP: Use More OOP

extracted ui class
extracted storage class
extracted parser class
extracted taskList class
extracted command class
extracted DukeException class

Commit:
Level-8: Cates and Times

Commit:
Level-8(v2.0): Dates and Times

displayed the time of deadlines and events in a different way from the way user entered them.

Commit:
A-JavaDoc

yuqiaoluolong committed:
Merge pull request #1 from yuqiaoluolong/branch-Level-8-Dates-and-Times

Branch level 8 dates and times

Commit:
New branch-A-JavaDoc
Merge branch 'master' of https://github.com/yuqiaoluolong/ip into branch-A-JavaDoc

# Conflicts:
#	src/main/java/Duke/command/Command.java
#	src/main/java/Duke/storage/Storage.java
#	src/main/java/Duke/task/Deadline.java
#	src/main/java/Duke/task/Event.java

Commit:
merge branch-A-JavaDoc and master

yuqiaoluolong committed:
Set theme jekyll-theme-merlot

Commit:
final version

Commit:
Merge branch 'master' of https://github.com/yuqiaoluolong/ip

Commit:
Duke: final version

Commit:
Changed the path so that the jar file can work correctly.

Commit:
Changed the file path so that Duke can store the task list into a file whose path is the same with that of Duke.jar

Commit:
Changed the icon so that it can be shown crossing platforms
